pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,378 @@
+		return reconcile.Result{}, nil
+	}
+	if installation.Spec.CNI == nil || installation.Spec.CNI.Type == "" {
+		r.status.SetDegraded(operator.ResourceNotReady, "Waiting for CNI type to be configured on Installation", nil, reqLogger)
+		return reconcile.Result{}, nil
+	}
+
+	// If the installation is terminating, do nothing.
+	if installation.DeletionTimestamp != nil {
+		reqLogger.Info("Installation is terminating, skipping reconciliation")
+		return reconcile.Result{}, nil
+	}
+
+	// Get all IP pools currently in the cluster.
+	currentPools := &crdv1.IPPoolList{}
+	err := r.client.List(ctx, currentPools)
+	if err != nil && !errors.IsNotFound(err) {
+		r.status.SetDegraded(operator.ResourceReadError, "error querying IP pools", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Write default IP pool configuration back to the Installation object using patch.
+	preDefaultPatchFrom := client.MergeFrom(installation.DeepCopy())
+	if err = fillDefaults(ctx, r.client, installation, currentPools); err != nil {
+		r.status.SetDegraded(operator.ResourceReadError, "error filling IP pool defaults", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+	if err = ValidatePools(installation); err != nil {
+		r.status.SetDegraded(operator.InvalidConfigurationError, "error validating IP pool configuration", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+	if err := r.client.Patch(ctx, installation, preDefaultPatchFrom); err != nil {
+		r.status.SetDegraded(operator.ResourceUpdateError, "Failed to write defaults", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+	reqLogger.V(1).Info("Reconciling IP pools for installation", "installation", installation.Spec)
+
+	// Get the APIServer. If healthy, we'll use the projectcalico.org/v3 API for managing pools.
+	// Otherwise, we'll use the internal v1 API for bootstrapping the cluster until the API server is available.
+	// This controller will never delete pools using the v1 API, as deletion is a complex process and is only
+	// properly handled when using the v3 API.
+	apiserver, _, err := utils.GetAPIServer(ctx, r.client)
+	if err != nil && !apierrors.IsNotFound(err) {
+		r.status.SetDegraded(operator.ResourceNotReady, "Error querying APIServer", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+	apiAvailable := apiserver != nil && apiserver.Status.State == v1.TigeraStatusReady
+
+	// Create a lookup map of pools owned by this controller for easy access.
+	// This controller will ignore any IP pools that it itself did not create.
+	ourPools := map[string]crdv1.IPPool{}
+	for _, p := range currentPools.Items {
+		if hasOwner(&p, installation) {
+			// This pool is owned by the Installation object, so consider it ours.
+			ourPools[p.Spec.CIDR] = p
+		} else if p.Name == "default-ipv4-ippool" || p.Name == "default-ipv6-ippool" {
+			// For legacy installs prior to the existence of this controller, IP pools created by
+			// the operator would have these names. Assume they are ours.
```

Comment:
I think this is a dangerous assumption. In this case I think we should compare what is in the Installation resource to the IPPool resource and if they match then we can add the reference and manage the pool from then on.
We could even extend this for any pool and that could be the mechanism for taking an externally managed pool and allowing the operator to take it over.

Reply:
> Could you expand on this?

I was thinking the operator could take ownership of any IPPool resource that matches the spec of any pool specified in the Installation resource.

> I am not sure how to do that in a more "safe" way. I suppose you might just mean adding additional checks? i.e., if the name matches one of these two and the spec fields match?

Yes that is exactly what I meant.

Reply:
> I was thinking the operator could take ownership of any IPPool resource that matches the spec of any pool specified in the Installation resource.

This is probably the way to go, although it does add a little bit of extra work to the defaulting logic. Namely, older Installations may have the IP pool specified but no "name" field, since that is introduced in this PR. We will need to, as part of defaulting, set the "Name" field based on any matching IP pools we find within the cluster.

We probably only want to do this for "default-ipvx-ippool" though - anyone adding new pools to their Installation can fill in the name explicitly so we don't need to guess. This is tricky though, because we need to default the "name" field since it's optional and hasn't existed forever... so perhaps our defaulting _should_ look at existing IP pools as a first attempt at defaulting the name.

Reply:
Hah, turns out I did already implement defaulting IP pool names based on whether the CIDR matches... go figure! That's probably good:

- Defaulting updates any Installation pools with a name based on in cluster pools if they exist, followed by a generated name.
- We assume any IP pools in the cluster that match exactly the IP pools in the Installation are owned by the operator.

This logic has the upside of more easily allowing users to opt-in to operator ownership of their IP pools by simply adding them to the Installation. 

The downside is that it's a bit fuzzy and might be a bit awkward to implement - if you don't match the IP pool exactly, the operator will be confused and attempt to create a second IP pool with the same CIDR as an existing one, resulting in an error. Probably OK - it's easily recoverable and doesn't harm the cluster.